packages/SwingSet/src/kernel/virtualObjectManager.js

```diff
@@ -0,0 +1,342 @@
+    return JSON.parse(syscall.vatstoreGet(instanceKey));
+  }
+
+  /**
+   * Write an object's state to secondary storage.
+   *
+   * @param {string} instanceKey  The instance ID of the object whose state is
+   *    being stored.
+   * @param {*} rawData  A data object representing the state to be written.
+   */
+  function store(instanceKey, rawData) {
+    syscall.vatstoreSet(instanceKey, JSON.stringify(rawData));
+  }
+
+  /**
+   * Make a new kind of virtual object.
+   *
+   * @param {*} instanceMaker  A function of the form `instanceMaker(state)` that
+   *    will return a representative instance wrapped around the given state.
+   *
+   * @returns {*} a maker function that can be called to manufacture new
+   *    instance of this kind of object.  The parameters of the maker function
+   *    are those of the `initialize` method implemented in the representative
+   *    produced by the `instanceMaker` parameter.
+   */
+  function makeKind(instanceMaker) {
+    const kindID = `${allocateExportID()}`;
+    let nextInstanceID = 1;
+
+    function makeInnerSelf(instanceKey, rawData) {
+      const innerSelf = {
+        prev: undefined,
+        next: undefined,
+        instanceKey,
+        rawData,
+        dropData() {
+          store(instanceKey, innerSelf.rawData);
+          innerSelf.rawData = null;
+        },
+      };
+      return innerSelf;
+    }
+
+    function makeRepresentative(innerSelf, initArgs) {
+      function ensureState() {
+        if (!innerSelf.rawData) {
+          const instanceKey = innerSelf.instanceKey;
+          innerSelf = cache.lookup(instanceKey);
+          if (!innerSelf) {
+            innerSelf = makeInnerSelf(instanceKey, fetch(instanceKey));
+            cache.remember(innerSelf);
+          }
+        }
+        cache.refresh(innerSelf);
+      }
+
+      function wrapData() {
+        innerSelf.activeData = {};
+        for (const prop of Object.getOwnPropertyNames(innerSelf.rawData)) {
+          Object.defineProperty(innerSelf.activeData, prop, {
+            get: () => {
+              ensureState();
+              return m.unserialize(innerSelf.rawData[prop]);
+            },
+            set: value => {
+              ensureState();
+              innerSelf.rawData[prop] = m.serialize(value);
+            },
+          });
+        }
+        return harden(innerSelf.activeData);
+      }
+
+      if (!innerSelf.rawData && initArgs) {
+        innerSelf.rawData = {};
+        const tempInstance = instanceMaker(innerSelf.rawData);
+        tempInstance.initialize(...initArgs);
+        for (const prop of Object.getOwnPropertyNames(innerSelf.rawData)) {
+          innerSelf.rawData[prop] = m.serialize(innerSelf.rawData[prop]);
+        }
+      }
+      if (!innerSelf.activeData) {
+        wrapData();
+      }
+      const rawRepresentative = instanceMaker(innerSelf.activeData);
+      delete rawRepresentative.initialize;
+      const representative = harden(rawRepresentative);
```

Comment:
`harden()` strictly always returns its argument, so there's no need for two separate names here.

Reply:
This was more about expressing intent, as at the beginning the code is handling the object in its mutable persona while later it's dealing with the object in its immutable persona.  It's the same object, but in a different ontological status, hence the different names.  But it's an easy tweak if you have a strong opinion about this.

Reply:
I guess someone who doesn't entirely know how `harden` works could see `rep = harden(rawRep)` and suddenly doubt their understanding, like "wait, is it important that we not return `rawRepresentative`? is `rawRepresentative` still mutable??".  Since `harden()` is something you *do* to an object, I think calling it on `representative` would indicate the intent pretty well.

(I like the grammatical pattern: `harden(x)` vs a hypothetical `y = hardened(x)`, to indicate side-effect vs not. Python does the same: `list.sort()` sorts it in-place, `list.sorted()` returns a new sorted list).

Reply:
Done.